Firestore/Source/Remote/FSTDatastore.mm

```diff
@@ -245,16 +248,17 @@ - (void)commitMutations:(NSArray<FSTMutation *> *)mutations
   [self invokeRPCWithFactory:rpcFactory errorHandler:completion];
 }
 
-- (void)lookupDocuments:(NSArray<FSTDocumentKey *> *)keys
+- (void)lookupDocuments:(const std::vector<DocumentKey> &)keys
              completion:(FSTVoidMaybeDocumentArrayErrorBlock)completion {
   GCFSBatchGetDocumentsRequest *request = [GCFSBatchGetDocumentsRequest message];
   request.database = [self.serializer encodedDatabaseID];
-  for (FSTDocumentKey *key in keys) {
+  for (const DocumentKey &key : keys) {
     [request.documentsArray addObject:[self.serializer encodedDocumentKey:key]];
   }
 
   __block FSTMaybeDocumentDictionary *results =
       [FSTMaybeDocumentDictionary maybeDocumentDictionary];
+  const std::vector<DocumentKey> inputKeys = keys;
```

Comment:
Why is this copy necessary?

Reply:
Without the copy, test has strange behavior (I'll post update once I get close to my laptop). My guess is that when block captures a local C++ reference, something goes wrong (I really don't know how); but capture a local C++ variable has no issue. Even though both are done by capturing by value I believe...

Reply:
FSTTransactionTests::testDeleteDocumentAndPatch would fail with the following msg if not copying the reference to a local variable.

2018-03-23 20:50:37.351134-0400 Firestore_Example[56268:5199634] 4.8.0 - [Firebase/Firestore][I-FST000001] RPC BatchGetDocuments returned headers (whitelisted): {
    date = "Sat, 24 Mar 2018 00:50:37 GMT";
    "x-google-backends" = "127.0.0.1:8887,::1:8788,tarantula.wat.corp.google.com:8081";
    "x-google-gfe-request-trace" = "tarantula.wat.corp.google.com:8081,::1:8788,tarantula.wat.corp.google.com:8081";
    "x-google-service" = "esf_datastore";
}
2018-03-23 20:50:37.356705-0400 Firestore_Example[56268:5199623] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: Mismatch in docs returned from document lookup.'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010bec812b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010b55cf41 objc_exception_throw + 48
	2   CoreFoundation                      0x000000010becd2f2 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x0000000107b35d69 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 193
	4   Firestore_Example                   0x0000000106f79833 __41-[FIRTransaction getDocument:completion:]_block_invoke + 499
	5   Firestore_Example                   0x000000010705477f __52-[FSTTransaction lookupDocumentsForKeys:completion:]_block_invoke + 1055
	6   Firestore_Example                   0x0000000106f84322 __43-[FSTDatastore lookupDocuments:completion:]_block_invoke_3 + 2082
	7   Firestore_Example                   0x0000000106f892c5 -[FSTDispatchQueue enterCheckedOperation:] + 517
	8   Firestore_Example                   0x0000000106f8981a __34-[FSTDispatchQueue dispatchAsync:]_block_invoke + 42
	9   libdispatch.dylib                   0x000000010c65f177 _dispatch_call_block_and_release + 12
	10  libdispatch.dylib                   0x000000010c6601ba _dispatch_client_callout + 8
	11  libdispatch.dylib                   0x000000010c667d3d _dispatch_queue_serial_drain + 654
	12  libdispatch.dylib                   0x000000010c6685a0 _dispatch_queue_invoke + 329
	13  libdispatch.dylib                   0x000000010c664c97 _dispatch_queue_override_invoke + 477
	14  libdispatch.dylib                   0x000000010c66a980 _dispatch_root_queue_drain + 568
	15  libdispatch.dylib                   0x000000010c66a6ea _dispatch_worker_thread3 + 119
	16  libsystem_pthread.dylib             0x000000010cb231ca _pthread_wqthread + 1387
	17  libsystem_pthread.dylib             0x000000010cb22c4d start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb)

Reply:
This is the textbook case of something that needs a comment. Any random passerby might see this and think, "aha, I can save a copy!" without realizing that the copy is required for correctness.

Meanwhile the problem you're solving here is that this block is escaping the calling stack frame. I'm surprised it merely fails the assertion rather crashing outright.

Objective-C blocks only capture by value, and capturing the const reference of the argument by value doesn't do what we want. Capturing as you've done makes two copies which also isn't great (once to create `inputKeys` in the caller, once to create the captured version in the closure).

There are two reasonable solutions:
  * Mark `inputKeys` as a `__block` variable: this will put the initial copy directly into the heap-allocated closure
  * Make `inputKeys` a `std::shared_ptr<std::vector<DocumentKey>>`

The second option is slightly better because it is the solution we have to use in C++11 so it ports more directly.

Taken together this means this line could be:

```c++
// Allow keys to outlive this stack frame
std::shared_ptr<std::vector<DocumentKey>> inputKeys = std::make_shared(keys);
```

Note that if you ever need more than one thing to outlive the calling stack frame you can declare a closure struct to cut down on lots of individual `shared_ptr`s, with something like this:
```c++
struct Closure {
  std::vector<DocumentKey> keys;
  FSTMaybeDocumentDictionary* results;
};
```
We can avoid copying altogether if we make the parameter of type `std::vector<DocumentKey>` and then `std::move(keys)` into the closure:
```objc
- (void)lookupDocuments:(const std::vector<DocumentKey> &)keys
             completion:(FSTVoidMaybeDocumentArrayErrorBlock)completion {
  // ...
  struct Closure {
    std::vector<DocumentKey> keys;
    FSTMaybeDocumentDictionary* results;
  };

  std::shared_ptr<Closure> closure = std::make_shared<Closure>();
  closure->keys = std::move(keys);
  closure->results = [FSTMaybeDocumentDictionary maybeDocumentDictionary];

  // Now pass closure by value into the block (or lambda) ...
}
```

Reply:
done